cranelift/codegen/src/egraph.rs

```diff
@@ -0,0 +1,411 @@
+
+        // For each block in RPO, create an enode for block entry, for
+        // each block param, and for each instruction.
+        for &block in self.domtree.cfg_postorder().iter().rev() {
+            let loop_level = self.loop_analysis.loop_level(block);
+            let blockparam_start =
+                u32::try_from(self.blockparam_ids_tys.len()).expect("Overflow in blockparam count");
+            for (i, &value) in func.dfg.block_params(block).iter().enumerate() {
+                let ty = func.dfg.value_type(value);
+                let param = self
+                    .egraph
+                    .add(
+                        Node::Param {
+                            block,
+                            index: u32::try_from(i).expect("More than 2^32 blockparams"),
+                            ty,
+                            loop_level,
+                        },
+                        &mut self.node_ctx,
+                    )
+                    .get();
+                value_to_id.insert(value, param);
+                self.blockparam_ids_tys.push((param, ty));
+                self.stats.node_created += 1;
+                self.stats.node_param += 1;
+            }
+            let blockparam_end =
+                u32::try_from(self.blockparam_ids_tys.len()).expect("Overflow in blockparam count");
+            self.blockparams[block] = blockparam_start..blockparam_end;
+
+            let side_effect_start =
+                u32::try_from(self.side_effect_ids.len()).expect("Overflow in side-effect count");
+            for inst in func.layout.block_insts(block) {
+                let side_effect = has_side_effect(func, inst)
+                    || (func.dfg[inst].opcode().can_load()
+                        && !func.dfg[inst].memflags().unwrap().readonly())
+                    || func.dfg[inst].opcode().can_store();
+
+                // Build args from SSA values.
+                let args = EntityList::from_iter(
+                    func.dfg.inst_args(inst).iter().map(|&arg| {
+                        let arg = func.dfg.resolve_aliases(arg);
+                        *value_to_id
+                            .get(&arg)
+                            .expect("Must have seen def before this use")
+                    }),
+                    &mut self.node_ctx.args,
+                );
+
+                let results = func.dfg.inst_results(inst);
+
+                let types = self
+                    .node_ctx
+                    .types
+                    .from_iter(results.iter().map(|&val| func.dfg.value_type(val)));
+                let types = types.freeze(&mut self.node_ctx.types);
+
+                let mem_state = self.alias_analysis.get_state_for_load(inst);
+                let is_readonly_load = func.dfg[inst].opcode() == Opcode::Load && {
+                    let memflags = func.dfg[inst].memflags().unwrap();
+                    memflags.readonly() && memflags.notrap()
+                };
```

Comment:
> I'm guessing that, unlike in the alias analysis, here it really isn't important that the opcode be exactly Opcode::Load. Sign- or zero-extending loads and vector loads should also be okay to treat as pure if they're both readonly and notrap, right?

In theory yeah, that's absolutely true. However making that change is a little more than `op.can_load()` because we want exactly those loads that are "normal" (take a value from memory with no special semantics or other side-effects), and `.can_load()` is a much broader predicate that includes e.g. atomics and CAS; and there's no "normal load but maybe extending" predicate. So I wanted to avoid untested, uncertain complexity.

Happy to address this in a followup if we find cases where it's important (or just in general), though!

Reply:
(The suggestion is great though, applying that; thanks!)